<a href="https://colab.research.google.com/github/petrus1210/pemograman-dasar/blob/main/ATS_PRAKTIKUM_KELOMPOK_1.ipynb" target="_parent"><img src="https://colab.research.google.com/assets/colab-badge.svg" alt="Open In Colab"/></a>

In [ ]:
Nama Kelompok 1

-Rifa Halwani (4232211015)
-Petrus Randanu (4232411007)
-Adi Candra Edita (4232411022)

In [ ]:
#SOAL NO 2
def hitung_efisiensi(input_energi, output_energi):
    """
    Fungsi ini menghitung efisiensi turbin.
    Parameter:
        input_energi (float): Energi yang masuk ke sistem (Joule, kJ, dll)
        output_energi (float): Energi keluaran dari sistem (Joule, kJ, dll)
    Return:
        float: Nilai efisiensi (antara 0 - 1)
    """

    if input_energi == 0:
        print("Error: Input energi tidak boleh nol.")
        return None

    efisiensi = output_energi / input_energi

    print("=== Perhitungan Efisiensi Turbin ===")
    print(f"Input Energi   : {input_energi}")
    print(f"Output Energi  : {output_energi}")
    print(f"Efisiensi Turbin: {efisiensi:.2f} atau {efisiensi * 100:.1f}%")

    return efisiensi


hasil = hitung_efisiensi(500, 400)

if hasil is not None:
    print(f"\nNilai efisiensi yang dihitung adalah {hasil:.2f} atau {hasil * 100:.1f}%")

=== Perhitungan Efisiensi Turbin ===
Input Energi   : 500
Output Energi  : 400
Efisiensi Turbin: 0.80 atau 80.0%

Nilai efisiensi yang dihitung adalah 0.80 atau 80.0%


1.Definisi Fungsi: Fungsi hitung_efisiensi didefinisikan, siap untuk digunakan.

2.Panggilan: Fungsi dipanggil dengan input_energi = 500 dan output_energi = 400.

3.Validasi: Sistem memastikan input (500) tidak nol.teks tebal

4.Perhitungan: Efisiensi dihitung: 400 / 500 = 0.8.

5.Cetak Output 1: Rincian perhitungan dicetak: "Efisiensi Turbin: 0.80 atau 80.0%".

6.Kembalian Nilai: Nilai 0.8 dikembalikan dan disimpan di variabel hasil.

7.Cetak Output 2: Program utama mencetak hasil akhir: "Nilai efisiensi yang dihitung adalah 0.80 atau 80.0%".

In [ ]:
# SOAL NO 4
def sistem_katup(tekanan_uap, status_manual_override):
    """
    Fungsi ini mensimulasikan logika kerja katup pengaman (safety valve).

    Parameter:
        tekanan_uap (float): nilai tekanan uap (dalam psi)
        status_manual_override (bool): status override manual
                                       True  = katup dikunci manual (tidak boleh terbuka)
                                       False = sistem otomatis aktif (boleh terbuka)

    Return:
        str: status katup ('Terbuka' atau 'Tertutup')
    """

    print("=== SISTEM PEMANTAUAN KATUP PENGAMAN ===")
    print(f"Tekanan Uap Saat Ini     : {tekanan_uap} psi")
    print(f"Status Manual Override   : {status_manual_override}")
    print("-------------------------------------------")

    if tekanan_uap > 150 and status_manual_override == False:
        print("Kondisi terpenuhi:")
        print(" - Tekanan melebihi ambang batas 150 psi")
        print(" - Override manual TIDAK aktif (False)")
        print(">>> Hasil: Katup Terbuka untuk melepaskan tekanan berlebih.")
        status_katup = "Terbuka"
    else:
        print("Kondisi tidak terpenuhi:")
        if tekanan_uap <= 150:
            print(" - Tekanan masih di bawah batas aman (<= 150 psi).")
        if status_manual_override == True:
            print(" - Override manual aktif, sistem dikunci oleh operator.")
        print(">>> Hasil: Katup Tetap Tertutup demi keamanan.")
        status_katup = "Tertutup"

    print("===========================================\n")
    return status_katup



hasil1 = sistem_katup(160, False)

hasil2 = sistem_katup(120, False)

hasil3 = sistem_katup(180, True)

print("=== RANGKUMAN HASIL PENGUJIAN ===")
print(f"Skenario 1 (160 psi, override False): Katup {hasil1}")
print(f"Skenario 2 (120 psi, override False): Katup {hasil2}")
print(f"Skenario 3 (180 psi, override True ): Katup {hasil3}")
print("===================================")

=== SISTEM PEMANTAUAN KATUP PENGAMAN ===
Tekanan Uap Saat Ini     : 160 psi
Status Manual Override   : False
-------------------------------------------
Kondisi terpenuhi:
 - Tekanan melebihi ambang batas 150 psi
 - Override manual TIDAK aktif (False)
>>> Hasil: Katup Terbuka untuk melepaskan tekanan berlebih.

=== SISTEM PEMANTAUAN KATUP PENGAMAN ===
Tekanan Uap Saat Ini     : 120 psi
Status Manual Override   : False
-------------------------------------------
Kondisi tidak terpenuhi:
 - Tekanan masih di bawah batas aman (<= 150 psi).
>>> Hasil: Katup Tetap Tertutup demi keamanan.

=== SISTEM PEMANTAUAN KATUP PENGAMAN ===
Tekanan Uap Saat Ini     : 180 psi
Status Manual Override   : True
-------------------------------------------
Kondisi tidak terpenuhi:
 - Override manual aktif, sistem dikunci oleh operator.
>>> Hasil: Katup Tetap Tertutup demi keamanan.

=== RANGKUMAN HASIL PENGUJIAN ===
Skenario 1 (160 psi, override False): Katup Terbuka
Skenario 2 (120 psi, override False): Katup

Penjelasan Lengkap:

def sistem_katup(...): Fungsi utama untuk mensimulasikan sistem katup pengaman.

Docstring ("""...""") Memberi dokumentasi tentang parameter dan fungsinya, agar mudah dipahami oleh teknisi lain.

Bagian print(...) di awal fungsi Menampilkan data kondisi nyata dari sistem sebelum keputusan dibuat:

Tekanan uap saat ini

Status override manual

if tekanan_uap > 150 and status_manual_override == False: Kondisi majemuk menggunakan operator AND. Kedua kondisi harus benar agar katup boleh terbuka.

Bagian else: Menangani semua kondisi selain itu — misalnya tekanan belum cukup tinggi, atau override aktif.

Percabangan tambahan di dalam else: Memberi tahu alasan spesifik mengapa katup tidak terbuka:

Tekanan belum cukup tinggi

Override manual aktif

Tiga contoh pengujian (test case):

Tekanan tinggi, override nonaktif → katup terbuka

Tekanan rendah, override nonaktif → tetap tertutup

Tekanan tinggi, override aktif → tertutup

In [ ]:
# SOAL NO 6
tekanan_uap = 160.0
status_manual_override = False

if tekanan_uap > 150 and status_manual_override == False:
    print("Katup terbuka")
else:
    print("Katup tertutup")


Katup terbuka


Penjelasan singkat:

Operator and digunakan karena kedua kondisi harus terpenuhi.

Kondisi pertama: tekanan_uap > 150

Kondisi kedua: status_manual_override == False

Jika keduanya benar, maka program mencetak "Katup terbuka", selain itu "Katup tertutup".

In [ ]:
#SOAL NO 8

def hitung_efisiensi_termal(energi_keluar, energi_masuk):
    efisiensi = (energi_keluar / energi_masuk) * 100
    print(f"Efisiensi termal = {efisiensi:.2f}%")

# Contoh pemanggilan fungsi
hitung_efisiensi_termal(500, 1200)

Efisiensi termal = 41.67%


Penjelasan langkah kerja secara singkat: Soal NO 8: Efisiensi Termal

1.Panggilan Fungsi: Fungsi hitung_efisiensi_termal dipanggil dengan \Energi Keluar = 500 dan Energi Masuk = 1200.

2.Perhitungan: Efisiensi dihitung: (500 / 1200) \times 100 \approx 41.67.

3.Output: Nilai efisiensi dicetak sebagai persentase: Efisiensi termal = 41.67%.

In [ ]:
#SOAL NO 10

volume = 0

while volume < 100:
    volume += 10
    if volume == 80:
        print(f"WARNING: Hampir penuh! {volume} liter.")
    elif volume == 100:
        print(f"TANGKI PENUH - Pengisian berhenti. {volume} liter.")
    else:
        print(f"Pengisian: {volume} liter")


Pengisian: 10 liter
Pengisian: 20 liter
Pengisian: 30 liter
Pengisian: 40 liter
Pengisian: 50 liter
Pengisian: 60 liter
Pengisian: 70 liter
Pengisian: 90 liter
TANGKI PENUH - Pengisian berhenti. 100 liter.


Penjelasan langkah kerja singkat Soal NO 10: Simulasi Pengisian

1.Inisialisasi: volume dimulai dari 0.

2.Looping (10 L s/d 70 L): Loop while berjalan 7 kali. volume bertambah 10 dan mencetak status "Pengisian: X liter" (7 output).

3.Peringatan (80 L): volume mencapai 80. Kondisi if volume == 80 terpenuhi. Output: "WARNING: Hampir penuh! saat 80 liter".

4.Lanjutan (90 L): volume bertambah menjadi 90. Kondisi else terpenuhi. Output: "Pengisian: 90 liter".

5.Penuh (100 L): volume bertambah menjadi 100. Kondisi elif volume == 100 terpenuhi. Output: "TANGKI PENUH - Pengisian berhenti. saat 100 liter".

6.Selesai: Loop berhenti karena 100.

In [12]:
#chellange


HARGA_KWH_STANDAR = 1500  # Harga per kWh dalam Rupiah
BATAS_BAIK = 150          # Batas konsumsi (kWh) untuk kategori "Baik"
BATAS_WAJAR = 250         # Batas konsumsi (kWh) untuk kategori "Wajar"
JUMLAH_HARI = 30          # Asumsi jumlah hari dalam sebulan

# --- 1. Fungsi Utama ---
def hitung_dan_analisis_energi(data_konsumsi_harian):
    """
    Menghitung total konsumsi, biaya, dan memberikan analisis.
    Mengimplementasikan perulangan 'for' dan percabangan 'if-elif-else'.
    """
    total_konsumsi = 0

    # Perulangan for: Menghitung total konsumsi dari data harian
    for konsumsi in data_konsumsi_harian:
        total_konsumsi += konsumsi

    # Menghitung total biaya
    total_biaya = total_konsumsi * HARGA_KWH_STANDAR

    # Percabangan if-elif-else: Menentukan kategori konsumsi dan saran
    kategori = ""
    saran = ""
    if total_konsumsi <= BATAS_BAIK:
        kategori = "BAIK"
        saran = "Selamat! Anda sangat hemat energi. Pertahankan kebiasaan ini."
    elif total_konsumsi <= BATAS_WAJAR:
        kategori = "WAJAR"
        saran = "Konsumsi Anda cukup efisien. Coba cari potensi penghematan kecil lainnya."
    else:
        kategori = "BOROS"
        saran = "Perlu perhatian! Konsumsi Anda tinggi. Pertimbangkan untuk mematikan perangkat yang tidak perlu."

    return total_konsumsi, total_biaya, kategori, saran

# --- 2. Fungsi untuk Input Pengguna ---
def ambil_input_konsumsi():
    """
    Mengambil input konsumsi harian dari pengguna.
    Mengimplementasikan perulangan 'while' dan fungsi input.
    """
    data_konsumsi = []
    hari_ke = 1

    # Perulangan while: Meminta input untuk setiap hari hingga JUMLAH_HARI terpenuhi
    while hari_ke <= JUMLAH_HARI:
        while True:
            try:
                # Fungsi input()
                konsumsi_hari = float(input(f"Masukkan konsumsi listrik hari ke-{hari_ke} (kWh): "))
                # Percabangan if: Memastikan nilai tidak negatif
                if konsumsi_hari >= 0:
                    data_konsumsi.append(konsumsi_hari)
                    break
                else:
                    print("**Input tidak valid. Konsumsi tidak bisa negatif.**")
            except ValueError:
                print("**Input tidak valid. Masukkan angka.**")

        hari_ke += 1

    return data_konsumsi

# --- 3. Fungsi Utama Program (Main Function) ---
def main_program():
    """
    Fungsi utama untuk menjalankan alur program.
    """
    print("--- ⚡️ ANALISIS KONSUMSI ENERGI BULANAN ⚡️ ---")
    print(f"Harga standar per kWh: Rp {HARGA_KWH_STANDAR:,.0f}")
    print(f"Silakan masukkan data konsumsi listrik Anda selama {JUMLAH_HARI} hari.")
    print("-" * 40)

    # Memanggil Fungsi untuk input
    data_harian = ambil_input_konsumsi()

    # Memanggil Fungsi untuk perhitungan dan analisis
    total_konsumsi, total_biaya, kategori, saran = hitung_dan_analisis_energi(data_harian)

    print("-" * 40)
    print("✅ **HASIL ANALISIS ENERGI**")
    print(f"Total Konsumsi Bulanan ({JUMLAH_HARI} hari): **{total_konsumsi:.2f} kWh**")
    print(f"Estimasi Total Biaya: **Rp {total_biaya:,.0f}**")
    print(f"Kategori Penggunaan Energi: **{kategori}**")
    print("-" * 40)
    print(f"Saran: {saran}")

# Menjalankan program utama
if __name__ == "__main__":
    main_program()

--- ⚡️ ANALISIS KONSUMSI ENERGI BULANAN ⚡️ ---
Harga standar per kWh: Rp 1,500
Silakan masukkan data konsumsi listrik Anda selama 30 hari.
----------------------------------------
Masukkan konsumsi listrik hari ke-1 (kWh): 20
Masukkan konsumsi listrik hari ke-2 (kWh): 21
Masukkan konsumsi listrik hari ke-3 (kWh): 22
Masukkan konsumsi listrik hari ke-4 (kWh): 23
Masukkan konsumsi listrik hari ke-5 (kWh): 24
Masukkan konsumsi listrik hari ke-6 (kWh): 25
Masukkan konsumsi listrik hari ke-7 (kWh): 26
Masukkan konsumsi listrik hari ke-8 (kWh): 21
Masukkan konsumsi listrik hari ke-9 (kWh): 22
Masukkan konsumsi listrik hari ke-10 (kWh): 26
Masukkan konsumsi listrik hari ke-11 (kWh): 27
Masukkan konsumsi listrik hari ke-12 (kWh): 20
Masukkan konsumsi listrik hari ke-13 (kWh): 21
Masukkan konsumsi listrik hari ke-14 (kWh): 24
Masukkan konsumsi listrik hari ke-15 (kWh): 23
Masukkan konsumsi listrik hari ke-16 (kWh): 21
Masukkan konsumsi listrik hari ke-17 (kWh): 20
Masukkan konsumsi listrik hari

Fungsi (def) :Tiga fungsi utama digunakan: hitung_dan_analisis_energi(), ambil_input_konsumsi(), dan main_program(). Ini membagi kode menjadi blok-blok logis dan reusable.

Perulangan (for) :Digunakan di dalam fungsi hitung_dan_analisis_energi() untuk secara iteratif menjumlahkan semua nilai konsumsi harian di dalam list data_konsumsi_harian.

Perulangan (while) :Digunakan di dalam fungsi ambil_input_konsumsi() untuk terus meminta input dari pengguna (hari ke-1 sampai hari ke-30) dan juga untuk memastikan input yang dimasukkan valid (perulangan while True dengan try-except dan break).

Percabangan (if-elif-else) :Digunakan di dalam fungsi hitung_dan_analisis_energi() untuk menentukan kategori konsumsi energi (BAIK, WAJAR, atau BOROS) berdasarkan batas yang ditetapkan (BATAS_BAIK dan BATAS_WAJAR).

Percabangan if (Sederhana) :Digunakan di dalam perulangan while di ambil_input_konsumsi() untuk memvalidasi bahwa konsumsi energi yang dimasukkan tidak negatif.

Fungsi Lain :Fungsi bawaan Python seperti input(), print(), float(), dan try-except (untuk error handling input) juga digunakan untuk membuat program lebih interaktif dan tangguh terhadap error pengguna
